# Lab 3: Data Types and Arrays
Welcome to Lab 3!

So far, we've used Python to manipulate numbers and work with tables. But we need to discuss data types to deepen our understanding of how to work with data in Python.

In this lab, you'll first see how to represent and manipulate another fundamental type of data: text.  A piece of text is called a *string* in Python. You'll also see how to work with *arrays* of data. An array could contain all the numbers between 0 and 100 or all the words in the chapter of a book. Lastly, you'll create tables and practice analyzing them with your knowledge of table operations. The [Python Reference](http://data8.org/sp22/python-reference.html) has information that will be useful for this lab.



Set up the tests and imports by running the cell below.

In [2]:
# Just run this cell
import numpy as np
import math
from datascience import *

# 1. Text
Programming doesn't just concern numbers. Text is one of the most common data types used in programs. 

Text is represented by a **string value** in Python. The word "string" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [4]:
print("I <3", 'Data Science')

I <3 Data Science


Just as names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [6]:
one = 'two'
plus = '*'
print(one, plus, one)

two * two


**Question 1.1.** Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [10]:
woman_asking = 'The woman asked:' #store the woman asking's action as a variable
woman_quote = '"Can it be that you have come from outer space?"'#store the woman's dialogue as a variable
gagarin_reply = 'Gagarin replied:' #store the gagarin's response as a variable
gagarin_quote = '"As a matter of fact, I have!"' #store the gagarin's quote as a variable

print(woman_asking, woman_quote) #print the woman's action and quote
print(gagarin_reply, gagarin_quote) #print gagarin's reply and quote

The woman asked: "Can it be that you have come from outer space?"
Gagarin replied: "As a matter of fact, I have!"


END SOLUTION

## 1.1. String Methods

Strings can be transformed using **methods**. Recall that methods and functions are not the same thing. Here is the textbook section on string methods: [4.2.1 String Methods](https://inferentialthinking.com/chapters/04/2/1/String_Methods.html?highlight=methods). 

Here's a sketch of how to call methods on a string:

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)
    
One example of a string method is `replace`, which replaces all instances of some part of the original string (or a *substring*) with a new string. 

    <original string>.replace(<old substring>, <new substring>)
    
`replace` returns (evaluates to) a new string, leaving the original string unchanged.
    
Try to predict the output of this example, then run the cell!

In [12]:
# Replace one letter
bag = 'bag'
print(bag.replace('g', 't'), bag)

bat bag


You can call functions on the results of other functions.  For example, `max(abs(-5), abs(3))` evaluates to 5.  Similarly, you can call methods on the results of other method or function calls.

You may have already noticed one difference between functions and methods - a function like `max` does not require a `.` before it's called, but a string method like `replace` does. Here's a handy [Python reference](http://data8.org/fa21/python-reference.html). It's a good idea to refer to this whenever you're unsure of how to call a function or method.

In [14]:
# Calling replace on the output of another call to replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

**Question 1.1.1.** Use `replace` to transform the string `'hitchhiker'` into `'matchmaker'`. Assign your result to `new_word`.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [16]:
new_word = 'hitchhiker'.replace('hi', 'ma') #replace the hi in hitchhiker with ma
new_word #show new word

'matchmaker'

END SOLUTION

There are many more string methods in Python, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation. [Stack Overflow](http://stackoverflow.com) has a huge database of answered questions that often demonstrate how to use these methods to achieve various ends. Material covered in these resources that haven't already been introduced in this class will be out of scope. 

## 1.2. Converting to and from Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

In [18]:
8 + "8"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

However, there are built-in functions to convert numbers to strings and strings to numbers. Some of these built-in functions have restrictions on the type of argument they take:

|Function |Description|
|-|-|
|`int`|Converts a string of digits or a float to an integer ("int") value|
|`float`|Converts a string of digits (perhaps with a decimal point) or an int to a decimal ("float") value|
|`str`|Converts any value to a string|

Try to predict what data type and value `example` evaluates to, then run the cell.

In [20]:
example = 8 + int("10") + float("8")

print(example)
print("This example returned a " + str(type(example)) + "!")

26.0
This example returned a <class 'float'>!


Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

**Question 1.2.1.** Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [24]:
# Some text data:
one_year = "1618"
another_year = "1648"
diff=int(another_year)-int(one_year) #subtract the integer of another_year minus one_year
diff #print solution

30

END SOLUTION

## 1.3. Passing strings to functions

String values, like numbers, can be arguments to functions and can be returned by functions. 

The function `len` (derived from the word "length") takes a single string as its argument and returns the number of characters (including spaces) in the string.

Note that it doesn't count *words*. `len("one small step for man")` evaluates to 22 characters, not 5 words.

**Question 1.3.1.**  Use `len` to find the number of characters in the long string in the next cell.  Characters include things like spaces and punctuation. Assign `sentence_length` to that number.

(The string is the first sentence of the English translation of the French [Declaration of the Rights of Man](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  


Type your solution in the following cell. Be sure to include comments explaining your code.

In [26]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length=len(a_very_long_sentence) #find the len value of the sentence
sentence_length #show result

896

END SOLUTION

# 2. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying **each number** in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**. 

## 2.1. Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [28]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

**Question 2.1.1.** Make an array containing the numbers 0, 1, -1, and $\pi$, in that order.  Name it `interesting_numbers`.  

*Hint:* How did you get the value $\pi$ in lab 2?  You can refer to it the same way here. The `math` module has been imported at the top of this notebook.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [36]:
#Your comments and solution
interesting_numbers=make_array(0, -1, 1, math.pi) #store values in an array
print(interesting_numbers) #show result

[ 0.         -1.          1.          3.14159265]


END SOLUTION

**Question 2.1.2.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [38]:
hello_world_components = make_array("Hello", ",", " ", "world", "!") #store components in an array
print(hello_world_components) #print result

['Hello' ',' ' ' 'world' '!']


END SOLUTION

###  `np.arange`
Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Run the following cells to see some examples!

In [40]:
# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [42]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

**Question 2.1.3.** Import `numpy` as `np` and then use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)


Type your solution in the following cell. Be sure to include comments explaining your code.

In [46]:
import numpy as np #import numpy
multiples_of_99 = np.arange(0, 10000, 99) #store start, stop, and step values
print(multiples_of_99) #print the array

[   0   99  198  297  396  495  594  693  792  891  990 1089 1188 1287 1386
 1485 1584 1683 1782 1881 1980 2079 2178 2277 2376 2475 2574 2673 2772 2871
 2970 3069 3168 3267 3366 3465 3564 3663 3762 3861 3960 4059 4158 4257 4356
 4455 4554 4653 4752 4851 4950 5049 5148 5247 5346 5445 5544 5643 5742 5841
 5940 6039 6138 6237 6336 6435 6534 6633 6732 6831 6930 7029 7128 7227 7326
 7425 7524 7623 7722 7821 7920 8019 8118 8217 8316 8415 8514 8613 8712 8811
 8910 9009 9108 9207 9306 9405 9504 9603 9702 9801 9900 9999]


END SOLUTION

## 2.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population_amounts` that includes estimated world populations of every year from **1950** to roughly the present.  (The estimates come from the US Census Bureau website.)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn how to do that later in this lab!

In [23]:
import numpy as np
from datascience import *
population_amounts = Table.read_table("/Users/camel/Downloads/world_population.csv")
population_amounts

Index,Position,Year,Population
0,Item #1,1950,2557628654
1,Item #2,1951,2594939877
2,Item #3,1952,2636772306
3,Item #4,1953,2682053389
4,Item #5,1954,2730228104
5,Item #6,1955,2782098943
6,Item #7,1956,2835299673
7,Item #8,1957,2891349717
8,Item #9,1958,2948137248
9,Item #10,1959,3000716593


Here's how we get the first element of `population_amounts`, which is the world population in the first year in the dataset, 1950.

In [33]:
first_ele = population_amounts.row(0)
print(first_ele)

Row(Index=0, Position='Item #1', Year=1950, Population=2557628654)


The value of that expression is the number 2,557,628,654 (around 2.5 billion), because that's the first thing in the array `population_amounts`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is a weird convention in computer science.  **0 is called the *index* of the first item.**  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population_amounts`.  Read and run each cell.

In [35]:
# The 13th element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population_amounts.row(12)
population_1962

Row(Index=12, Position='Item #13', Year=1962, Population=3140093217)

In [37]:
# The 66th element is the population in 2015.
population_2015 = population_amounts.row(65)
population_2015

Row(Index=65, Position='Item #66', Year=2015, Population=7256490011)

In [39]:
# The array has only 66 elements, so this doesn't work.
# (There's no element with 66 other elements before it.)
population_2016 = population_amounts.row(66)
population_2016

IndexError: index 66 is out of bounds for axis 0 with size 66

Since `make_array` returns an array, we can call `.item(3)` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [41]:
make_array(-1, -3, 4, -2).item(3)

-2

**Question 2.2.1.** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population_amounts` using `item`.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [43]:
#I'm using row commands instead of item commands here because my Jupyter notebook wasn't working, hope that's alright.
population_1973 = population_amounts.row(23) #calls 23rd row since 1963-1950 is 23
population_1973 #print solution

Row(Index=23, Position='Item #24', Year=1973, Population=3942096442)

END SOLUTION

## 2.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Rounding
Here is one simple question we might ask about world population:

> How big was the population in each year, rounded to the nearest million?

Rounding is often used with large numbers when we don't need as much precision in our numbers. One example of this is when we present data in tables and visualizations. 

We could try to answer our question using the `round` function that is built into Python and the `item` method you just saw. 

**Note:** the `round` function takes in two arguments: the number to be rounded, and the number of decimal places to round to. The second argument can be thought of as how many steps right or left you move from the decimal point. Negative numbers tell us to move left, and positive numbers tell us to move right. So, if we have `round(1234.5, -2)`, it means that we should move two places left, and then make all numbers to the right of this place zeroes. This would output the number 1200.0. On the other hand, if we have `round(6.789, 1)`, we should move one place right, and then make all numbers to the right of this place zeroes. This would output the number 6.8.

In [85]:
population_1950_magnitude = round(population_amounts.item(0), -6)
population_1951_magnitude = round(population_amounts.item(1), -6)
population_1952_magnitude = round(population_amounts.item(2), -6)
population_1953_magnitude = round(population_amounts.item(3), -6)

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `round` that rounds each element of an array.  It takes in two arguments: a single array of numbers, and the number of decimal places to round to.  It returns an array of the same length, where the first element of the result is the first element of the argument rounded, and so on.

**Question 2.3.1.** Use `np.round` to compute the world population in every year, rounded to the nearest million (6 zeroes).  Give the result (an array of 66 numbers) the name `population_rounded`.  Your code should be very short.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [49]:
#Your comments and solution
the_world_pop = Table.read_table("/Users/camel/Downloads/world_population.csv") #calls the file
population_rounded = np.round(the_world_pop.column('Population') / 1e6) #rounds every number to 10 to the power of 6 (6 zeroes)
print(population_rounded) #print the array

[ 2558.  2595.  2637.  2682.  2730.  2782.  2835.  2891.  2948.  3001.
  3043.  3084.  3140.  3210.  3281.  3350.  3421.  3490.  3562.  3637.
  3713.  3790.  3867.  3942.  4017.  4089.  4160.  4232.  4304.  4379.
  4451.  4534.  4615.  4696.  4775.  4856.  4941.  5027.  5115.  5201.
  5289.  5372.  5456.  5538.  5619.  5699.  5779.  5858.  5935.  6012.
  6089.  6165.  6242.  6319.  6396.  6473.  6551.  6630.  6709.  6788.
  6866.  6944.  7022.  7101.  7179.  7256.]


END SOLUTION

What you just did is called **elementwise** application of `np.round`, since `np.round` operates separately on each element of the array that it's called on. 

##### Arithmetic
Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [55]:
population_amounts = Table.read_table("/Users/camel/Downloads/world_population.csv") #import table
population_in_billions = population_amounts.with_columns('Population in Billions', population_amounts.column('Population') / 1e9) #add new column
print(population_in_billions) #print array

Index | Position | Year | Population | Population in Billions
0     | Item #1  | 1950 | 2557628654 | 2.55763
1     | Item #2  | 1951 | 2594939877 | 2.59494
2     | Item #3  | 1952 | 2636772306 | 2.63677
3     | Item #4  | 1953 | 2682053389 | 2.68205
4     | Item #5  | 1954 | 2730228104 | 2.73023
5     | Item #6  | 1955 | 2782098943 | 2.7821
6     | Item #7  | 1956 | 2835299673 | 2.8353
7     | Item #8  | 1957 | 2891349717 | 2.89135
8     | Item #9  | 1958 | 2948137248 | 2.94814
9     | Item #10 | 1959 | 3000716593 | 3.00072
... (56 rows omitted)


You can do the same with addition, subtraction, multiplication, and exponentiation (`**`).


## 3. Creating Tables

An array is useful for describing a single attribute of each element in a collection. For example, let's say our collection is all US States. Then an array could describe the land area of each state. 

Tables extend this idea by containing multiple columns stored and represented as arrays, each one describing a different attribute for every element of a collection. In this way, tables allow us to not only store data about many entities but to also contain several kinds of data about each entity. 

For example, in the cell below we have two arrays. The first one, `population_amounts`, was defined above in section 2.2 and contains the world population in each year (estimated by the US Census Bureau). The second array, `years`, contains the years themselves. These elements are in order, so the year and the world population for that year have the same index in their corresponding arrays.

In [65]:
# Just run this cell
population_amounts = Table.read_table("/Users/camel/Downloads/world_population.csv") #import table
years = np.arange(1950, 2015+1)
print("Population column:", population_amounts)
print("Years column:", years)

Population column: Index | Position | Year | Population
0     | Item #1  | 1950 | 2557628654
1     | Item #2  | 1951 | 2594939877
2     | Item #3  | 1952 | 2636772306
3     | Item #4  | 1953 | 2682053389
4     | Item #5  | 1954 | 2730228104
5     | Item #6  | 1955 | 2782098943
6     | Item #7  | 1956 | 2835299673
7     | Item #8  | 1957 | 2891349717
8     | Item #9  | 1958 | 2948137248
9     | Item #10 | 1959 | 3000716593
... (56 rows omitted)
Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963 1964
 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977 1978 1979
 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994
 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009
 2010 2011 2012 2013 2014 2015]


Suppose we want to answer this question:

> In which year did the world's population cross 6 billion?

You could technically answer this question just from staring at the arrays, but it's a bit convoluted, since you would have to count the position where the population first crossed 6 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a *`Table`*, a 2-dimensional type of dataset. 

The expression below:

- creates an empty table using the expression `Table()`,
- adds two columns by calling `with_columns` with four arguments,
- assigns the result to the name `population`, and finally
- evaluates `population` so that we can see the table.

The strings `"Year"` and `"Population"` are column labels that we have chosen. The names `population_amounts` and `years` were assigned above to two arrays of the **same length**. The function `with_columns` (you can find the documentation [here](http://data8.org/fa21/python-reference.html)) takes in alternating strings (to represent column labels) and arrays (representing the data in those columns). The strings and arrays are separated by commas.

In [69]:
from datascience import Table
population_column = population_amounts.column('Population')
population = Table().with_columns(
    "Population", population_column,
    "Year", years
)
print(population)

Population | Year
2557628654 | 1950
2594939877 | 1951
2636772306 | 1952
2682053389 | 1953
2730228104 | 1954
2782098943 | 1955
2835299673 | 1956
2891349717 | 1957
2948137248 | 1958
3000716593 | 1959
... (56 rows omitted)


Now the data is combined into a single table! It's much easier to parse this data. If you need to know what the population was in 1959, for example, you can tell from a single glance.

**Question 3.1.** In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a table that has two columns called "Name" and "Rating", which hold `top_10_movie_names` and `top_10_movie_ratings` respectively.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [71]:
top_10_movie_names = make_array(
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)')
top_10_movie_ratings = make_array(9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8)
top_10_movies = Table().with_columns(
    "Name", top_10_movie_names,
    "Rating", top_10_movie_ratings
) #create the table
print(top_10_movies) #print table

Name                                                     | Rating
The Shawshank Redemption (1994)                          | 9.2
The Godfather (1972)                                     | 9.2
The Godfather: Part II (1974)                            | 9
Pulp Fiction (1994)                                      | 8.9
Schindler's List (1993)                                  | 8.9
The Lord of the Rings: The Return of the King (2003)     | 8.9
12 Angry Men (1957)                                      | 8.9
The Dark Knight (2008)                                   | 8.9
Il buono, il brutto, il cattivo (1966)                   | 8.9
The Lord of the Rings: The Fellowship of the Ring (2001) | 8.8


END SOLUTION

#### Loading a table from a file



In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we load them in from an external source, like a data file. There are many formats for data files, but CSV ("comma-separated values") is the most common.

`Table.read_table(...)` takes one argument (a path to a data file in **string** format) and returns a table.  

**Question 3.2.** `imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

(You may remember working with this table in Lab 2!)


Type your solution in the following cell. Be sure to include comments explaining your code.

In [73]:
#Your comments and solution
imdb = Table.read_table("/Users/camel/Downloads/imdb.csv") #import table
imdb #show table

Votes,Rating,Title,Year,Decade
88355,8.4,M,1931,1930
132823,8.3,Singin' in the Rain,1952,1950
74178,8.3,All About Eve,1950,1950
635139,8.6,Léon,1994,1990
145514,8.2,The Elephant Man,1980,1980
425461,8.3,Full Metal Jacket,1987,1980
441174,8.1,Gone Girl,2014,2010
850601,8.3,Batman Begins,2005,2000
37664,8.2,Judgment at Nuremberg,1961,1960
46987,8,Relatos salvajes,2014,2010


END SOLUTION

Where did `imdb.csv` come from? Take a look at [this lab's folder](./). You should see a file called `imdb.csv`.

Open up the `imdb.csv` file in that folder and look at the format. What do you notice? The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

## 4. More Table Operations!

Now that you've worked with arrays, let's add a few more methods to the list of table operations that you saw in Lab 2.

### `column`

`column` takes the column name of a table (in string format) as its argument and returns the values in that column as an **array**. 

In [75]:
# Returns an array of movie names
top_10_movies.column('Name')

array(['The Shawshank Redemption (1994)', 'The Godfather (1972)',
       'The Godfather: Part II (1974)', 'Pulp Fiction (1994)',
       "Schindler's List (1993)",
       'The Lord of the Rings: The Return of the King (2003)',
       '12 Angry Men (1957)', 'The Dark Knight (2008)',
       'Il buono, il brutto, il cattivo (1966)',
       'The Lord of the Rings: The Fellowship of the Ring (2001)'],
      dtype='<U56')

### `take`
The table method `take` takes as its argument an array of numbers.  Each number should be the index of a row in the table.  It returns a **new table** with only those rows. 

You'll usually want to use `take` in conjunction with `np.arange` to take the first few rows of a table.

In [77]:
# Take first 5 movies of top_10_movies
top_10_movies.take(np.arange(0, 5, 1))

Name,Rating
The Shawshank Redemption (1994),9.2
The Godfather (1972),9.2
The Godfather: Part II (1974),9
Pulp Fiction (1994),8.9
Schindler's List (1993),8.9


The next three questions will give you practice with combining the operations you've learned in this lab and the previous one to answer questions about the `population` and `imdb` tables. First, check out the `population` table from section 2.

In [79]:
# Run this cell to display the population table.
population

Population,Year
2557628654,1950
2594939877,1951
2636772306,1952
2682053389,1953
2730228104,1954
2782098943,1955
2835299673,1956
2891349717,1957
2948137248,1958
3000716593,1959


**Question 4.1.** Check out the `population` table from section 2 of this lab.  Compute the year when the world population first went above 6 billion. Assign the year to `year_population_crossed_6_billion`.


Type your solution in the following cell. Be sure to include comments explaining your code.# BEGIN SOLUTION

In [87]:
#Your comments and solution
year_population_crossed_6_billion = population.where('Population', are.above(6e9)).column('Year').item(0) #find what year the population went over 6 bil
print(year_population_crossed_6_billion) #print result

1999


END SOLUTION

**Question 4.2.** Find the average rating for movies released before the year 2000 and the average rating for movies released in the year 2000 or after for the movies in `imdb`.

*Hint*: Think of the steps you need to do (take the average, find the ratings, find movies released in 20th/21st centuries), and try to put them in an order that makes sense.


Type your solution in the following cell. Be sure to include comments explaining your code.# BEGIN SOLUTION

In [93]:
#Your comments and solution
before_2000 = imdb.where('Year', are.below(2000)) #find movies released before 2000
after_or_2000 = imdb.where('Year', are.above_or_equal_to(2000)) #find movies released after 2000

average_rating_before_2000 = before_2000.column('Rating').mean() #find average rating for movies before 2000
average_rating_after_or_2000 = after_or_2000.column('Rating').mean() #find average rating for movies after 2000

print("Average before 2000 rating:", before_2000) #print result
print("Average after or in 2000 rating:", after_or_2000) #print result

Average before 2000 rating: Votes  | Rating | Title                 | Year | Decade
88355  | 8.4    | M                     | 1931 | 1930
132823 | 8.3    | Singin' in the Rain   | 1952 | 1950
74178  | 8.3    | All About Eve         | 1950 | 1950
635139 | 8.6    | Léon                  | 1994 | 1990
145514 | 8.2    | The Elephant Man      | 1980 | 1980
425461 | 8.3    | Full Metal Jacket     | 1987 | 1980
37664  | 8.2    | Judgment at Nuremberg | 1961 | 1960
56842  | 8.1    | It Happened One Night | 1934 | 1930
156432 | 8.3    | Some Like It Hot      | 1959 | 1950
264285 | 8.1    | The Princess Bride    | 1987 | 1980
... (161 rows omitted)
Average after or in 2000 rating: Votes  | Rating | Title                                | Year | Decade
441174 | 8.1    | Gone Girl                            | 2014 | 2010
850601 | 8.3    | Batman Begins                        | 2005 | 2000
46987  | 8      | Relatos salvajes                     | 2014 | 2010
502773 | 8.2    | Gran Torino             

END SOLUTION

**Question 4.3.** Here's a challenge: Find the number of movies that came out in *even* years.

*Hint:* The operator `%` computes the remainder when dividing by a number.  So `5 % 2` is 1 and `6 % 2` is 0.  A number is even if the remainder is 0 when you divide by 2.

*Hint 2:* `%` can be used on arrays, operating elementwise like `+` or `*`.  So `make_array(5, 6, 7) % 2` is `array([1, 0, 1])`.

*Hint 3:* Add a new column called "Year Remainder" that's the remainder when each movie's release year is divided by 2. To do this, you can use `tbl.with_column(col_name, col_values)`: a table method that takes in the name of the new column and an array of values and returns a copy of the original `tbl` with the new column.  Then, use `where` to find rows where that new column is equal to 0.  Finally, use `num_rows` to count the number of such rows.

*Note:* These steps can be chained in one single statement, or broken up across several lines with intermediate names assigned. You’re always welcome to break down problems however you wish!


Type your solution in the following cell. Be sure to include comments explaining your code.

In [97]:
imdb_with_remainder = imdb.with_column(
    'Year Remainder', imdb.column('Year') % 2
) #adds a new column that contains the remainder when the year is 0

even_years = imdb_with_remainder.where('Year Remainder', 0) #get rid of rows where remainder is zero

num_even_year_movies = even_years.num_rows #find movies released in even years

print(num_even_year_movies)

127


END SOLUTION

## 5. More Array Practice

The following questions will be great practice for learning how to generate arrays in different ways, as well as executing different methods of array arithmetic!

###  5.1 `np.arange` (cont.)

##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Oakland, California site for the month of December 2015.  To analyze the data, we want to know when each reading was taken, but we find that the data don't include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2015 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 5.1.1.** Create an array of the *time, in seconds, since the start of the month* at which **each hourly reading** was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too!  If your `collection_times` isn't passing the tests, check its length and make sure it has $31 \times 24$ elements.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [101]:
import numpy as np
num_readings = 31 * 24 #hours in december
collection_times = np.arange(0, num_readings * 3600, 3600) #create array of times in seconds
print(collection_times) #print result
len(collection_times) #show length

[      0    3600    7200   10800   14400   18000   21600   25200   28800
   32400   36000   39600   43200   46800   50400   54000   57600   61200
   64800   68400   72000   75600   79200   82800   86400   90000   93600
   97200  100800  104400  108000  111600  115200  118800  122400  126000
  129600  133200  136800  140400  144000  147600  151200  154800  158400
  162000  165600  169200  172800  176400  180000  183600  187200  190800
  194400  198000  201600  205200  208800  212400  216000  219600  223200
  226800  230400  234000  237600  241200  244800  248400  252000  255600
  259200  262800  266400  270000  273600  277200  280800  284400  288000
  291600  295200  298800  302400  306000  309600  313200  316800  320400
  324000  327600  331200  334800  338400  342000  345600  349200  352800
  356400  360000  363600  367200  370800  374400  378000  381600  385200
  388800  392400  396000  399600  403200  406800  410400  414000  417600
  421200  424800  428400  432000  435600  439200  4

744

END SOLUTION

### 5.2 Doing something to every element of an array (cont.)

Calculate a tip on several restaurant bills at once (in this case just 3):

In [103]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [ 20.12  39.9   31.01]
Tips:			 [ 4.024  7.98   6.202]



**Question 5.2.1** Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [109]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
total_charges = restaurant_bills * 1.2
print(total_charges)

[ 24.144  47.88   37.212]


END SOLUTION

**Question 5.2.2.** The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one in `more_restaurant_bills`.  How is your code different?


Type your solution in the following cell. Be sure to include comments explaining your code.

In [129]:
more_restaurant_bills = Table.read_table("/Users/camel/Downloads/more_restaurant_bills.csv").column("Bill") #import file
more_total_charges = more_restaurant_bills * 1.2 #multiply numbers by 1.2
print(more_total_charges[:10]) #Show the first 10 charges

[ 20.244  20.892  12.216  11.808  22.128  10.788  47.664   4.26    5.58
   6.732]


 END SOLUTION

The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 5.2.3.** What was the sum of all the bills in `more_restaurant_bills`, *including tips*?


Type your solution in the following cell. Be sure to include comments explaining your code.

In [131]:
total_charges = more_restaurant_bills * 1.2 #compute total charge of all bills including the 1.2 tip
sum_of_bills = total_charges.sum() #find sum of all charges
print(sum_of_bills) #print result

1795730.064


END SOLUTION

**Question 5.2.4.** The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

*Hint 1:* `np.arange(1, 2**30, 1)` creates an array with $2^{30}$ elements and **will crash your kernel**.

*Hint 2:* Part of your solution will involve `np.arange`, but your array shouldn't have more than 30 elements.


Type your solution in the following cell. Be sure to include comments explaining your code.

In [ ]:
import numpy as np
exponents = np.arange(30)
powers_of_2 = 2 ** e
powers_of_2

END SOLUTION

You're done with Lab 2! 